### PR TITLE
Add DVC server, and ainput channel to server

### PR DESCRIPTION
**Pull request by elmarco**

### CODE REVIEW

**Review by CBenoit — COMMENTED** [review]
Thank you for this work!

To make sure we are on the same page, turning the methods of the SvcProcessor trait into async is a very controversial change that will be hard to merge.
We already had quite lengthy conversations about that with @pacmancoder and @ibeckermayer , and we concluded it was better for these to _not_ be async.

There are many considerations to keep in mind.
One of the biggest problem I remember is that processing a given channel was causing everything else to hang.

I can see this code is also suffering from this:

```rust
active_stage.process(&mut image, action, &payload).await?
```

This `await` operator is invoked inside the main loop, nested into a `tokio::select!`.
The problem is that it may take an arbitrary amount of time to return, and this is entirely dependent on the channel implementation.
While we await, the loop is stopped and we stop processing new events from both the client and the server.

Fixing this requires taking other controversial decisions with even deeper consequences.

In short, I’m afraid we can’t take this route, even if it could be convenient.

The approach we decided to take is to spawn a platform-dependent "backend" on the side, and communicate using explicit message passing. The main example for that is the clipboard channel. The code for the processor itself is still portable which is nice because we can reuse it for all the platforms, including WASM in the browser, where the execution model is quite different.

**Comment by elmarco** [comment]
> To make sure we are on the same page, turning the methods of the SvcProcessor trait into async is a very controversial change that will be hard to merge.

Hm... then we may have to change the server traits to not be async either, and use event system. This is less convenient and perhaps less idiomatic. Tracking associated request/reply may be troublesome too..

**Comment by CBenoit** [comment]
> > To make sure we are on the same page, turning the methods of the SvcProcessor trait into async is a very controversial change that will be hard to merge.
> 
> Hm... then we may have to change the server traits to not be async either, and use event system. This is less convenient and perhaps less idiomatic. Tracking associated request/reply may be troublesome too..

If you can find a nice alternative using async, I would sure be happy to see it. However, I don’t think it’s evident that we’ll get a more idiomatic and ergonomic API just by adding async / await keywords, at least not without much deeper refactoring. Also, achieving both "idiomatic" and "unidiomatic" API is possible with both "blocking" and async Rust. All I want to convey is that it’s not a trivial change and it will take probably more efforts than expected to make it happen. (And we actually already spent a good amount of time into trying this approach in the past.)

**Comment by elmarco** [comment]
@CBenoit ok, so the current version should be ok, as I dropped async stuff

File: crates/ironrdp-dvc/src/server.rs [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+    let mut res = Vec::new();
+    for msg in messages {
+        let total_size = msg.size();
+
+        let msg = encode_vec(msg.as_ref())?;
+        let mut off = 0;
+
+        while off < total_size {
+            let rem = total_size.checked_sub(off).unwrap();
+            let size = core::cmp::min(rem, 1590);
```
Comment (pacmancoder) [inline crates/ironrdp-dvc/src/server.rs:234]: nit: move `1590` to named constant

File: crates/ironrdp-pdu/src/rdp/vc/dvc/ainput.rs [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+    pub time: u64,
+    pub flags: MouseEventFlags,
+    pub x: i32,
+    pub y: i32,
+}
+
+impl MousePdu {
+    const NAME: &'static str = "AInputMousePdu";
+
+    const FIXED_PART_SIZE: usize = core::mem::size_of::<u64>() * 2 + core::mem::size_of::<i32>() * 2;
```
Comment (pacmancoder) [inline crates/ironrdp-pdu/src/rdp/vc/dvc/ainput.rs:163]: (style) Our new code style states the following about size fields:

>Use an inline comment for each field of the structure.

```rust
// GOOD
const FIXED_PART_SIZE: usize = 1 /* Version */ + 1 /* Endianness */ + 2 /* CommonHeaderLength */ + 4 /* Filler */;

// BAD
const FIXED_PART_SIZE: usize = size_of::<u8>() + size_of::<u8>() + size_of::<u16>() + size_of::<u32>();
```

Please update all `FIXED_PART_SIZE` according to the new style

File: crates/ironrdp-dvc/src/server.rs [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+        vc::dvc::ClientPdu::from_buffer(&mut user_data, user_data_len).map_err(|e| custom_err!("DVC client PDU", e))?;
+
+    // … | DvcData ]
+    let dvc_data = user_data;
+
+    Ok(DynamicChannelCtx { dvc_pdu, dvc_data })
+}
+
+fn encode_dvc_message(pdu: vc::dvc::ServerPdu) -> PduResult<SvcMessage> {
+    // FIXME: use PduEncode instead
```
Comment (pacmancoder) [inline crates/ironrdp-dvc/src/server.rs:218]: Should this `FIXME` be removed or fixed?

File: crates/ironrdp-pdu/src/rdp/vc/dvc/ainput.rs [hunk 4]
```diff
@@ -0,0 +1,259 @@
+use bitflags::bitflags;
```
Comment (pacmancoder) [inline crates/ironrdp-pdu/src/rdp/vc/dvc/ainput.rs:1]: It would be great to add a link to FreeRDP docs for AInput PDUs structures

**Review by pacmancoder — COMMENTED** [review]
Nice work! I have some comments from my side, but nothing overly critical -- the only thing that I think may be improved is Arc<Mutex<_>> use for `RdpServerInputHandler`, maybe we could come up with something better?

File: crates/ironrdp-dvc/src/server.rs (hunk 1 above)
Comment (elmarco) [reply crates/ironrdp-dvc/src/server.rs:234]: ok

File: crates/ironrdp-pdu/src/rdp/vc/dvc/ainput.rs (hunk 2 above)
Comment (elmarco) [reply crates/ironrdp-pdu/src/rdp/vc/dvc/ainput.rs:163]: ok

File: crates/ironrdp-dvc/src/server.rs (hunk 3 above)
Comment (elmarco) [reply crates/ironrdp-dvc/src/server.rs:218]: I am working on it, but it's another branch.. tedious conversion from legacy to/from buffer to encode/decode trait...

File: crates/ironrdp-pdu/src/rdp/vc/dvc/ainput.rs (hunk 4 above)
Comment (elmarco) [reply crates/ironrdp-pdu/src/rdp/vc/dvc/ainput.rs:1]: ok

File: crates/ironrdp-server/src/server.rs [hunk 5]
```diff
@@ -40,6 +42,41 @@ impl RdpServerSecurity {
     }
 }
 
+pub struct AInputHandler {
```
[the remark this answers is not in the document]
Comment (elmarco) [reply crates/ironrdp-server/src/server.rs:45]: Actually, it shouldn't be public. Fixed. Fwiw, I think we may want to have implementation and traits for known channels, but that can come later - it depends what API and flexibilty is exposed by the server to some degree.

File: crates/ironrdp-server/src/server.rs [hunk 6]
```diff
@@ -102,7 +138,7 @@ impl RdpServerSecurity {
 /// ```
 pub struct RdpServer {
     opts: RdpServerOptions,
-    handler: Box<dyn RdpServerInputHandler>,
+    handler: Arc<Mutex<Box<dyn RdpServerInputHandler>>>,
```
[the remark this answers is not in the document]
Comment (elmarco) [reply crates/ironrdp-server/src/server.rs:141]: I don't have a good idea.. if we use a channel at this level, then maybe we should completely change the API and not use RdpServerInputHandler (& other) trait objects. But I would rather leave that choice to the implementation of the server at this point and prefer trait objects, but they need to be shareable via arc<mutex> then.

We can leave a TODO note?

[the remark this answers is not in the document]
Comment (CBenoit) [reply crates/ironrdp-server/src/server.rs:176]: If it’s too much work, I’m fine with the TODO note for now, and we can revisit later.

File: crates/ironrdp-pdu/src/rdp/vc/dvc.rs [hunk 7]
```diff
@@ -11,6 +11,7 @@ use crate::PduParsing;
 #[cfg(test)]
 mod tests;
 
+pub mod ainput;
```
Comment (CBenoit) [inline crates/ironrdp-pdu/src/rdp/vc/dvc.rs:14]: thought: Ideally, I would like to move code for specific channels in their own crates. For instance, we do this for the RDPDR static channel, for which the code is found in the `ironrdp-rdpdr` crate. In this case, a `ironrdp-ainput` crate could be created. It’s fine to do a follow-up PR for that though.

File: crates/ironrdp-pdu/src/lib.rs [hunk 8]
```diff
@@ -157,6 +157,25 @@ pub trait PduEncode {
...
+    /// Returns the associated PDU name associated.
+    fn name(&self) -> &'static str {
+        "legacy-pdu-encode"
+    }
+
+    /// Computes the size in bytes for this PDU.
+    fn size(&self) -> usize {
+        self.len()
+    }
+}
```
[the remark this answers is not in the document]
Comment (elmarco) [reply crates/ironrdp-pdu/src/lib.rs:177]: No, it's temporary. switching all the code to PduEncode is tedious.. I have other work pending, but enough in this (and other) MR at this point.

This is used in DVC server implementation, to wrap ServerPdu in SvcMessage (taking PduEncode).

[the remark this answers is not in the document]
Comment (elmarco) [reply crates/ironrdp-pdu/src/lib.rs:177]: moved to legacy

File: crates/ironrdp-pdu/src/rdp/vc/dvc.rs (hunk 7 above)
Comment (elmarco) [reply crates/ironrdp-pdu/src/rdp/vc/dvc.rs:14]: done

File: crates/ironrdp-dvc/src/server.rs [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+            processor: Box::new(processor),
+            data: CompleteData::new(),
+        }
+    }
+}
+/// DRDYNVC Static Virtual Channel (the Remote Desktop Protocol: Dynamic Virtual Channel Extension)
+///
+/// It adds support for dynamic virtual channels (DVC).
+pub struct DrdynvcServer {
+    dynamic_channels: Slab<DynamicChannel>,
```
[the remark this answers is not in the document]
Comment (elmarco) [reply crates/ironrdp-dvc/src/server.rs:57]: A slab returns the key when storing the value.  That let it manage the dynamically allocated channel ids. (and ids can be reused if a channel is replaced etc).

A vec or a btreemap or other collection don't let you do that conveniently. (Linux kernel has idr & xarray). The client doesn't need it here, has it doesn't have to do ID allocation.  

Basically slab is a convenient Vec<Option<T>> under the hood. 

slab is used by thousands of projects: https://lib.rs/crates/slab/rev

File: crates/ironrdp-server/src/server.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (elmarco) [reply crates/ironrdp-server/src/server.rs:176]: added TODO

**Comment by pacmancoder** [comment]
LGTM from my side! 🚀

**Review by pacmancoder — APPROVED** [review]
LGTM from my side! 🚀

**Review by CBenoit — APPROVED** [review]
LGTM! Great work, thank you again!
